R/case-variables.R

```diff
@@ -0,0 +1,57 @@
+#' @rdname describe
+#' @export
+setMethod("name", "Case", function (x) x@name)
+#' @rdname describe
+#' @export
+setMethod("name<-", "Case",
+          function (x, value) {
+              x@name <- validateNewName(value)
+              return(x)
+          })
+
+#' Make a case variable
+#' 
+#' The `makeCaseVariable` function derives a variable using values from other 
+#' variables. These are evaluated in the order they are supplied in the list 
+#' as the `cases` argument (they proceed in an IF, ELSE IF, ELSE IF, ..., ELSE 
+#' fashion); the first one that matches selects the corresponding value from 
+#' the [`Case`] object.
+#' 
+#' @param cases A list of objects of class [`Case`]
+#' @param ... other properties to pass about the case variable (e.g. name)
+#' @export
+makeCaseVariable <- function (cases, ...) {
```

Comment:
Looking at the code, I'm not convinced that the `Case` class helps anything. (And I realize that I'm probably responsible for sending you down that path.)  An S4 class might make more sense if there were some input constraints you could enforce in the constructor, "Cases" don't really have meaning or validity outside of the totality of the case statement--i.e. you can't have a default "id" for a Case generically. So then `Case()` could be done as `list()` with some validation inside the variable building function.  

How about something like this: 

    makeCaseVariable <- function (..., cases, else, name) {
        dots <- list(...)
        if (missing(cases)) {
            is_expr <- function (x) inherits(x, "CrunchLogicalExpression")
            exprs <- Filter(is_expr, dots)
            dots <- Filter(Negate(is_expr), dots)
            cases <- mapply(function (expr, name) list(name=name, expression=expr), 
                expr=exprs, name=names(exprs))
        }
        # Else (and continuing), `cases` is a list of list objects with minimum 
        # "name" and "expression".
        # Now, fill in the potentially omitted attributes (id, missing, etc.)
        # and validate them
        cases <- ensureValidCases(cases)
        ...
    }

If you supply "cases", it works similar to how `combine` does (like `cases=list(list(name="Young men", expression=ds$gender == "Male" & ds$age < 34), ...)`. Alternatively, the function provides a shorter version where you could do something like 

    makeCaseVariable(
        `Young men`=ds$gender == "Male" & ds$age < 34, 
        `Old men`=ds$gender == "Male" & ds$age >=35,
        else="Women",
        name="Men by age group",
        description="More words about this variable"
    )

when you don't care about numeric values and ids and such.